File: backend/java-api/src/main/java/com/zerocost/entity/Event.java
package com.zerocost.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.GenericGenerator;
import org.locationtech.jts.geom.Point;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "events", indexes = {
    @Index(name = "idx_events_start_time", columnList = "start_time"),
    @Index(name = "idx_events_category", columnList = "category_id"),
    @Index(name = "idx_events_created_at", columnList = "created_at")
})
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Event {

    @Id
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "latitude", nullable = false)
    private Double latitude;

    @Column(name = "longitude", nullable = false)
    private Double longitude;

    @Column(name = "location", columnDefinition = "geography(Point,4326)")
    private Point location;

    @Column(name = "address")
    private String address;

    @Column(name = "start_time", nullable = false)
    private Instant startTime;

    @Column(name = "end_time")
    private Instant endTime;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "source_id")
    private Source source;

    @Column(name = "source_url")
    private String sourceUrl;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "organizer_name")
    private String organizerName;

    @Column(name = "organizer_contact")
    private String organizerContact;

    @Column(name = "capacity")
    private Integer capacity;

    @Column(name = "is_verified")
    private Boolean isVerified = false;

    @Column(name = "raw_data", columnDefinition = "jsonb")
    private String rawData;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private Instant updatedAt;
}

File: backend/go-scraper/internal/models/event.go
package models

import "time"

// Event represents a scraped free event/opportunity
type Event struct {
	Title        string                 `json:"title"`
	Description  string                 `json:"description"`
	Latitude     float64                `json:"latitude"`
	Longitude    float64                `json:"longitude"`
	Address      string                 `json:"address,omitempty"`
	StartTime    time.Time              `json:"start_time"`
	EndTime      *time.Time             `json:"end_time,omitempty"`
	Category     string                 `json:"category"`
	Source       string                 `json:"source"`
	SourceURL    string                 `json:"source_url"`
	ImageURL     string                 `json:"image_url,omitempty"`
	Organizer    string                 `json:"organizer,omitempty"`
	Capacity     int                    `json:"capacity,omitempty"`
	RawData      map[string]interface{} `json:"raw_data,omitempty"`
}

// ScraperResult contains events scraped from a source
type ScraperResult struct {
	Source       string
	Events       []Event
	ErrorCount   int
	Duration     time.Duration
}


